store.go

```diff
@@ -0,0 +1,193 @@
+		buckets[i] = bucket{
+			entries: make(map[string]*entry, maxSize),
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &expirableStore{
+		maxSize:       maxSize,
+		items:         make(map[string]*entry, maxSize),
+		buckets:       buckets,
+		bucketTTL:     bucketTTL,
+		numberBuckets: opts.withNumberBuckets,
+		pool: sync.Pool{
+			New: func() any {
+				return &entry{
+					value: &Quota{},
+				}
+			},
+		},
+		cancelFunc: cancel,
+		ctx:        ctx,
+	}
+
+	go s.deleteExpired()
+	return s, nil
+}
+
+func (s *expirableStore) shutdown() error {
+	s.cancelFunc()
+	return nil
+}
+
+func (s *expirableStore) deleteExpired() {
+	ticker := time.NewTicker(s.bucketTTL)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker.C:
+			s.emptyExpiredBucket()
+		}
+	}
+}
+
+// TODO: document this
+func (s *expirableStore) fetch(id string, limit *Limit) (*Quota, error) {
+	select {
+	case <-s.ctx.Done():
+		return nil, ErrStopped
+	default:
+		// continue
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := getKey(limit.Resource, limit.Action, string(limit.Per), id)
+
+	e, ok := s.items[key]
+	switch {
+	case !ok:
+		e = s.pool.Get().(*entry)
+		e.key = key
+		e.value.reset(limit)
+		if err := s.add(e); err != nil {
+			s.pool.Put(e)
+			return nil, err
+		}
+		s.addToBucket(e)
+	case e.value.Expired():
+		s.removeFromBucket(e)
+		e.value.reset(limit)
+		s.addToBucket(e)
+	}
+
+	return e.value, nil
+}
+
+func (s *expirableStore) add(e *entry) error {
+	if _, ok := s.items[e.key]; !ok && len(s.items) >= s.maxSize {
+		return ErrLimiterFull
+	}
+	s.items[e.key] = e
+	return nil
+}
+
+func (s *expirableStore) addToBucket(e *entry) {
+	e.bucket = (int(e.value.limit.Period/s.bucketTTL) + s.nextBucketToExpire) % s.numberBuckets
+	s.buckets[e.bucket].entries[e.key] = e
+	if s.buckets[e.bucket].expiresAt.Before(e.value.expiresAt) {
+		s.buckets[e.bucket].expiresAt = e.value.expiresAt
+	}
+}
+
+// emptyExpiredBuckets is called via a go routine. It should run approximately
+// once every s.bucketTTL to delete all of the items in the next expired bucket.
+func (s *expirableStore) emptyExpiredBucket() {
+	s.mu.Lock()
+
+	toExpire := s.nextBucketToExpire
+	s.nextBucketToExpire = (s.nextBucketToExpire + 1) % s.numberBuckets
+
+	timeToExpire := time.Until(s.buckets[toExpire].expiresAt)
+	// Just in case, check to see if this has run early and there is still some
+	// time before the bucket expires. in which case wait until the bucket has
+	// expired before deleting.
+	if timeToExpire > 0 {
+		s.mu.Unlock()
+		time.Sleep(timeToExpire)
+		s.mu.Lock()
+	}
+	defer s.mu.Unlock()
+	for _, delEnt := range s.buckets[toExpire].entries {
+		s.removeElement(delEnt)
+	}
+}
+
+func (s *expirableStore) removeElement(e *entry) {
+	delete(s.items, e.key)
+	s.removeFromBucket(e)
+	s.pool.Put(e)
+}
+
+func (s *expirableStore) removeFromBucket(e *entry) {
+	delete(s.buckets[e.bucket].entries, e.key)
```

Comment:
looks racey.

Reply:
And this is called from either `fetch` or via `emptyExpiredBucket` which will grab a lock.